src/main/java/com/google/devtools/build/lib/syntax/MethodDescriptor.java

```diff
@@ -59,19 +62,157 @@ private MethodDescriptor(
+        // because their true receiver is StringModule.INSTANCE.
+        mh = MethodHandles.insertArguments(mh, 0, StringModule.INSTANCE);
+      }
+
+      if (!useStarlarkThread) {
+        mh = MethodHandles.dropArguments(mh, mh.type().parameterCount(), StarlarkThread.class);
+      }
+
+      mh = MethodHandleUtil.moveArgument(mh, mh.type().parameterCount() - 1, 1);
+
+      // mh: (self, StarlarkThread, args...) -> ?
+
+      mh = mh.asSpreader(Object[].class, mh.type().parameterCount() - 2);
+
+      // mh: (self, StarlarkThread, args) -> ?
+
+      if (mh.type().returnType() == void.class) {
+        mh =
+            MethodHandles.filterReturnValue(
+                mh, MethodHandles.constant(NoneType.class, Starlark.NONE));
+        mh = MethodHandles.dropArguments(mh, mh.type().parameterCount(), Mutability.class);
+      } else if (Starlark.validValueClass(mh.type().returnType())) {
+        mh = handleNullReturn(mh, allowReturnNones, name);
+        mh = MethodHandles.dropArguments(mh, mh.type().parameterCount(), Mutability.class);
+      } else {
+        mh = handleNullReturn(mh, allowReturnNones, name);
+
+        MethodHandle fromJava =
+            MethodHandles.explicitCastArguments(
+                FROM_JAVA,
+                MethodType.methodType(Object.class, mh.type().returnType(), Mutability.class));
+        mh = MethodHandles.collectArguments(fromJava, 0, mh);
+      }
+
+      mh = MethodHandleUtil.moveArgument(mh, 3, 2);
+
+      Class<?> expectedP0 = isStringModule ? String.class : method.getDeclaringClass();
+      // Self-check: better fail earlier.
+      Preconditions.checkState(mh.type().parameterCount() == 4, "%s", mh.type());
+      Preconditions.checkState(mh.type().parameterType(0) == expectedP0, "%s", mh.type());
+      Preconditions.checkState(mh.type().parameterType(1) == StarlarkThread.class, "%s", mh.type());
+      Preconditions.checkState(mh.type().parameterType(2) == Mutability.class, "%s", mh.type());
+      Preconditions.checkState(mh.type().parameterType(3) == Object[].class, "%s", mh.type());
+
+      mh =
+          MethodHandles.explicitCastArguments(
+              mh,
+              MethodType.methodType(
+                  Object.class,
+                  Object.class,
+                  StarlarkThread.class,
+                  Mutability.class,
+                  Object[].class));
+
+      this.methodHandle = mh;
+    } catch (Exception e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  private static MethodHandle handleNullReturn(
+      MethodHandle mh, boolean allowReturnNones, String name) {
+    if (mh.type().returnType().isPrimitive()) {
+      return mh;
+    }
+
+    if (allowReturnNones) {
+      // TODO(adonovan): eliminate allowReturnNones. Given that we convert
+      //   String/Integer/Boolean/List/Map, it seems obtuse to crash instead
+      //   of converting null too.
+      return nullToNone(mh);
+    } else {
+      return nullToException(mh, name);
```

Comment:
I'm curious what the stack looks like when this fails.

Reply:
Probably depends on JVM version, on my laptop it looks like these functions are called directly from `call`:

```
test test test
java.lang.AssertionError: 
	at com.google.devtools.build.lib.syntax.MethodDescriptor.nullToNone(MethodDescriptor.java:223)
	at com.google.devtools.build.lib.syntax.MethodDescriptor.call(MethodDescriptor.java:388)
	at com.google.devtools.build.lib.syntax.BuiltinCallable.fastcall(BuiltinCallable.java:73)
	at com.google.devtools.build.lib.syntax.Starlark.fastcall(Starlark.java:445)
	at com.google.devtools.build.lib.syntax.Eval.evalCall(Eval.java:610)
	at com.google.devtools.build.lib.syntax.Eval.eval(Eval.java:432)
```